### ISSUE TITLE
Termux API doesn't work on Android O

### ISSUE THREAD

**Issue by Jim-Zenn** [termux/termux-api#105]
I get 
`cmd: Can't find service: acitivity`
for the following commands:
`termux-battery-status`
`termux-clipboard-get`
`termux-clipboard-set`
`termux-camera-info`
`termux-contact-list`

I have installed Termux API and I am running Android 8.0.0 Build number OPP4.170623.014 on Google Pixel C.

**Comment by eladts** [comment]
I have the same issues with Nexus 5X running Android 8.0.0 build number OPR6.170623.013. All the commands from the termux-api package fail with the same message. I tried uninstalling and reinstalling Termux::API, but to no avail.

**Comment by datalin** [comment]
Same issue with Android O installed on Nexus 6p

**Comment by fornwall** [comment]
Hm, seems that calling `am` does not work at all on Android 8.0?

The script `/system/bin/am` is now:

```sh
if [ "$1" != "instrument" ] ; then
    cmd activity "$@"
else
    base=/system
    export CLASSPATH=$base/framework/am.jar
    exec app_process $base/bin com.android.commands.am.Am "$@"
fi
```

In Android 7.0 it was:

```sh
base=/system
export CLASSPATH=$base/framework/am.jar
exec app_process $base/bin com.android.commands.am.Am "$@"
```

**Comment by eladts** [comment]
Wouldn't it be simpler to simply integrate Termux::API into the main Termux app? What are the benefits of having Termux::API as a seperate app?

**Comment by sdrausty** [comment]
@eladts there are many benefits to keep Termux::API as an add-on. 
1. Permission issues
2. Modularity
I can probably think of more. Please keep add-ons separate @fornwall 

Add-on suggestion: integrating https://github.com/svn2github/ghostcommander as a Termux add-on. This would open up $HOME and similar from the Android interface.

**Comment by fornwall** [comment]
Right, the main reason for keeping the API package separate is that it needs a lot of permissions, which is nice to avoid requiring in the base app.

**Comment by assimilat** [comment]
I got the api to (seemingly) work on a rooted Pixel XL by modifying /data/data/com.termux/files/usr/bin/am to call /system/bin/cmd via tsu. I know this is a dirty hack(and script) I was in a bit of a rush...hope it might help somehow.
`#!/data/data/com.termux/files/usr/bin/bash
INPUT=$(echo "$@" | sed -e "s/^.*[-][-]user/--user/g");
COMMAND='LD_LIBRARY_PATH=/system/lib64 /system/bin/cmd activity broadcast';
tsu -c -s "LD_LIBRARY_PATH=/system/lib64 /system/bin/sh -c \"$COMMAND $INPUT\""`

**Comment by fornwall** [comment]
Link by `concludent` on #termux: https://pastebin.com/Las0Cv7e

**Comment by eladts** [comment]
While not part of termux-api the command termux-wake-lock also does not work on Oreo.

**Comment by y2kbugger** [comment]
+1

**Comment by Jim-Zenn** [comment]
@fornwall I am a bit confused. Is there a work around? Will this be fixed?

**Comment by brancengregory** [comment]
+1

**Comment by jonathanwalker** [comment]
+1

**Comment by gustavjg** [comment]
+1

**Comment by ofus** [comment]
+2

**Comment by fornwall** [comment]
@Jim-Zenn I'm looking into it. I was hoping that it would be possible to use `/system/bin/am` even on Android 8 with some workaround, but it seems we have to use another mechanism (a service listening to connections).

**Comment by moldabekov** [comment]
@fornwall any news? The only thing that keeps me from upgrading to Oreo is Termux. I heavily rely on termux-api in some of my apps. Since Termux is killer app, how can the community help you? Maybe should you publish list of issues you face by this time?

**Comment by CrisVM** [comment]
@fornwall Any news with the fix? I have already upgraded my OnePlus 3T to Oreo and have a few scripts that depend on Termux-API to work properly... I should have checked for problems like this prior to upgrading, but now it's too late... 😕  _(And now I don't feel like messing with some obscure procedure just to downgrade my phone..._ 😁 _)_
Thanks in advance.

**Comment by heyitswither** [comment]
+1

**Comment by Bestfast** [comment]
+1

**Comment by srajasimman** [comment]
Kindly give us a fix plz. My many scripts depends termux-api

**Comment by ofus** [comment]
@fornwall <https://github.com/fornwall> Is a workaround possible?  Do you
have the bandwidth & desire to work on one at some point, or is this
something that needs to be delegated/forked?

**Comment by MikeAmy** [comment]
Can anyone else work on it? I have the time but not the android experience.

Also, is having a separate process problematic? Processes limit on Android. There's really no easy API any more?

**Comment by sdrausty** [comment]
If Termux were to use the more current java, will this help in overcoming Android 8 issues?

**Comment by zingmars** [comment]
No, the problem has to do with Android changing their API in a way that breaks it. Nothing to do with the way termux is compiled.

**Comment by jzr** [comment]
I agree that Android has changed the API and this broke termux-api functionality.
However, I don't think it was broken on purpose.

I tried to follow the stack trace (and the analysis provided by concludent):

`runSendBroadcast` UNCONDITIONALLY provides `null` as a parameter (`IApplicationThread caller`) to `broadcastIntent`, causing an exception later in `isInstantApp`.

The reason it works when called with `su` on a rooted phone is that there is a special if-condition, that'll skip the code that throws an exception for any `UID < FIRST_APPLICATION_UID`.

The result of `isInstantApp` however, is only used once, for setting flags on the intent.

```
final boolean callerInstantApp = isInstantApp(callerApp, callerPackage, callingUid);
// Instant Apps cannot use FLAG_RECEIVER_VISIBLE_TO_INSTANT_APPS
if (callerInstantApp) {
    intent.setFlags(intent.getFlags() & ~Intent.FLAG_RECEIVER_VISIBLE_TO_INSTANT_APPS);
}
```
There are several other calls to `broadcastIntentLocked` that will run into the same exception, as they provide `null` for `callerApp` / `callerPackage`. 

Based on this, I'd argue that calling `broadcastIntent` in this fashion is intended to work (but doesn't), and the **bug** is in `isInstantApp`. It should handle `null`, instead of throwing an Exception.

Is anyone familiar with Android source / bug reporting that could confirm this and raise at a place where it may get discussed / resolved?

**Comment by zingmars** [comment]
You can find the bug trackers here - https://source.android.com/setup/report-bugs#bug-queues. I suppose this would go under system?

I don't know how quick their process is, however. Given it's technical nature it will probably take a while. 

That said, it's very likely that even if a patch does get deployed, it will take ages to actually reach people and there will be plenty of people with 8.0 that has this functionality broken :/

**Comment by sdrausty** [comment]
Where are the sources to the Android 8 API?

**Comment by fornwall** [comment]
Thanks to really nice work by @michalbednarski in https://github.com/termux/termux-packages/pull/2090 it seems that we may finally have a working am program in Termux even on Android 8!

Try installing the new termux-am package with `pkg install termux-am` and report how it works for you. This package will probably be added to the set of base packages that every Termux setup has installed in a short while if it works ok.

**Comment by CesarGA** [comment]
After installing `termux-am` and running `termux-camera-info` I'm getting the following error

```
LINK EXECUTABLE "/system/bin/dex2oat": "/data/data/com.termux/files/usr/lib/libtermux-exec.so" is 64-bit instead of 32-bit
```

`dpkg --print-architecture` returns `aarch64`

This is a Pixel 2 XL running Android 8.1.0

I have no idea about how to point to the 64 bit version, so any help is appreciated.

**Comment by thedward** [comment]
I've got a Nexus 5X running 8.1 and I just installed `termux-am`, now I'm getting the following error when I try to use the affected `termux-*` commands.

```
CANNOT LINK EXECUTABLE "/system/bin/dex2oat": "/data/data/com.termux/files/usr/lib/libtermux-exec.so" is 64-bit instead of 32-bit
```

And then it just hangs until I interrupt it.

**Comment by fornwall** [comment]
@CesarGA @thedward Thanks a lot! Can you update with `pkg up` and see if the updated version of termux-am works better?

**Comment by CesarGA** [comment]
@fornwall now it just hangs without the error.

**Comment by fornwall** [comment]
@CesarGA Thanks! Two possible problems:

1. You have the Termux:API app installed, right?
2. Are you running it directly on the device, not remotely from over ssh or similar (on first run you will get permission prompts for it to work, so you need to be doing that first run on-device)?

**Comment by thedward** [comment]
@formwall, it works! Yay! Many thanks to you and to @michalbednarski.

**Comment by brancengregory** [comment]
I'm getting the following error on install of termux-am on Pixel XL:

Err:1 https://termux.net stable/main all termux-am all 0.1
  404  Not Found
E: Failed to fetch
https://termux.net/dists/stable/main/binary-all/termux-am_0.1_all.deb  404
Not Found
E: Internal Error, ordering was unable to handle the media swap

Any ideas?

**Comment by brancengregory** [comment]
A restart after updating Termux:API did the trick. Thanks!!

**Comment by CrisVM** [comment]
Thanks!!! You guys rock!!!
It works perfectly on my OnePlus 3T now! :clap: :clap: :clap: :fireworks: :tada:

**Comment by MikeAmy** [comment]
Tested on Sony Experia XZ Premium running Android 8.0.0.

Just tested a few things so far.

Working:
termux-camera-info
termux-camera-photo
termux-location
termux-tts-speak
termux-battery-status

Tested termux-toast but nothong happened. What should happen?

Thanks for this it's really great!

**Comment by MikeAmy** [comment]
ok, strange, termux-toast is now working.

Btw, I did the pkg install termux-am. Nothing I tested worked. Then updated termux-api in play store and did pkg up termux-api inside termux. Then it started rocking. Cheers!

**Comment by tmzullinger** [comment]
Thanks @michalbednarski and @fornwall!  This works well on 8.1 on Pixel 2 XL (for my relatively limited testing).  Is there any particular testing that would be helpful?

**Comment by cscullion** [comment]
Maybe I missed a step. I updated Termux:API from google play.  Then I ran pkg up.  I also tried rebooting.  However, on my Pixel C running 8.1, I get this error when trying to use either termux-clipboard-set or termux-clipboard-get:

"execv("/data/data/com.termux/files/usr/bin/am"): No such file or directory"

**Comment by tmzullinger** [comment]
@cscullion, you also need to install termux-am: `pkg install termux-am`

**Comment by cscullion** [comment]
OK... I'm an idiot.  Thanks.  That did the trick.  I can report that both clipboard get and set are working on a Pixel C running 8.1.

**Comment by FiveYellowMice** [comment]
Awesome, just updated Termux-API from Google Play and installed termux-am from APT, and finally get them to work on HTC U11 (Android 8.0).

**Comment by moldabekov** [comment]
Good job guys!

I confirm that Termux-API works well on Oreo on OnePlus 5, except termux-notification. It doesn't send any notifications.

**Comment by regagain** [comment]
For me it just hangs with no error, using Termux and Termux API from F-Droid. 
Is that because F-Droid is stuck on Termux API v0.15?

**Comment by fornwall** [comment]
@regagain Try running `pkg up` to update and try again!

Thanks for reporting!

As of now termux-tools (which everyone has installed) depends on termux-am, so every existing user will have it pulled in as a dependency on the next update (and new ones will have termux-am preinstalled). This means that the `am` command, and thus the termux API commands, should work out of the box also on Android 8, so I'm finally closing this issue!

**Comment by deuxverres** [comment]
I have found some wonky behaviour since in my first 8.1 install the termux API and storage access didn't work, but after a ROM re-flash they just started working again...

**Comment by chaoqing** [comment]
`termux-notification` still do not work. It just exit without any action.

**Comment by moldabekov** [comment]
+1 for `termux-notification`. It is still doesn't work for me

**Comment by MMaster** [comment]
termux-notification definitely doesn't work on Oreo. Unfortunately while I can read code and know what needs to be done I don't have Android dev. environment setup so the best I can do right now is provide info on what needs to be done and hope that some good samaritan will do it.

After reviewing the termux-api code I found the issue. Android O update requires you to setup channel for notifications. I found out that NotificationAPI doesn't do this. 

Without going into too much detail and just repeating what someone else has already said, answer to this SO question explains very well what needs to be done in order to make it work again on Oreo:
https://stackoverflow.com/questions/43093260/notification-not-showing-in-android-8-oreo/44534559

Hope this helps.
